source/MaterialXGenShader/UnitSystem.cpp

```diff
@@ -54,24 +55,37 @@ void ScalarUnitNode::emitFunctionDefinition(const ShaderNode& node, GenContext&
     ScopedFloatFormatting fmt(Value::FloatFormatFixed, 15);
     unitLUT.add(Type::FLOATARRAY, VAR_UNIT_SCALE, Value::createValue<vector<float>>(unitScales));
 
-    BEGIN_SHADER_STAGE(stage, Stage::PIXEL)
     const ShaderGenerator& shadergen = context.getShaderGenerator();
-    shadergen.emitString("float mx_" + _scalarUnitConverter->getUnitType() + "_unit_ratio(int unit_from, int unit_to)", stage);
-    shadergen.emitLineBreak(stage);
+    shadergen.emitLine("float mx_" + _scalarUnitConverter->getUnitType() + "_unit_ratio(int unit_from, int unit_to)", stage, false);
     shadergen.emitScopeBegin(stage);
-
-    shadergen.emitLineBreak(stage);
     shadergen.emitVariableDeclarations(unitLUT, shadergen.getSyntax().getConstantQualifier(), ";", context, stage, true);
-
-    shadergen.emitLineBreak(stage);
-    shadergen.emitString("return ("+ VAR_UNIT_SCALE + "[unit_from] / " + VAR_UNIT_SCALE + "[unit_to]);", stage);
-    shadergen.emitLineBreak(stage);
+    shadergen.emitLine("return ("+ VAR_UNIT_SCALE + "[unit_from] / " + VAR_UNIT_SCALE + "[unit_to])", stage);
     shadergen.emitScopeEnd(stage);
     shadergen.emitLineBreak(stage);
-    END_SHADER_STAGE(shader, Stage::PIXEL)
+END_SHADER_STAGE(shader, Stage::PIXEL)
+}
+
+void ScalarUnitNode::emitFunctionCall(const ShaderNode& node, GenContext& context, ShaderStage& stage) const
+{
+BEGIN_SHADER_STAGE(stage, Stage::PIXEL)
+    const ShaderGenerator& shadergen = context.getShaderGenerator();
 
-    // Emit registered implementation
-    SourceCodeNode::emitFunctionDefinition(node, context, stage);
+    const ShaderInput* in = node.getInput(0);
+    const ShaderInput* from = node.getInput(1);
+    const ShaderInput* to = node.getInput(2);
+
+    shadergen.emitLineBegin(stage);
+    shadergen.emitOutput(node.getOutput(), true, false, context, stage);
```

Comment:
Since the unit conversion is just a float multiply we can inline this so the same code works for any input type (as long as the input type supports multiply by float).

Reply:
So let's say I have two inputs:
```
      <input name="realworld_scale" type="vector2" value="1, 1" unit="inch" unittype="distance"/>
      <input name="depth" type="float" value="0.2" value="0.125" unit="inch" unittype="distance"/>
```
With this approach the call is inlined as 
`float heightnormal_texturemap_depth_unit_out = heightnormal_texturemap_depth_unit_in * mx_distance_unit_ratio(heightnormal_texturemap_depth_unit_unit_from, u_distanceUnitTarget);` and 
`vec2 heightnormal_texturemap_realworld_scale_unit_out = heightnormal_texturemap_realworld_scale_unit_in * mx_distance_unit_ratio(heightnormal_texturemap_realworld_scale_unit_unit_from, u_distanceUnitTarget);`

But since `emitFunctionDefinition` is called for each type, we still have duplicate `float mx_distance_unit_ratio(int unit_from, int unit_to){...}`

Reply:
Another thing is that we've removed the flexibility of a scalar node to perform actions other than a multiply. A multiply + addition option is removed for instance.

Reply:
Sorry I forgot to set the hash correctly. Using the unit ratio function name as hash we can make sure the ratio function is only emitted once for every unit type. Change list has been updated to correct this now.

Reply:
Re: _"A multiply + addition option is removed for instance"_

I thought we decided to only support scale/multiply unit conversions? However, even if we want to add other conversion expressions later this design will not prevent us from doing that. Note that an arbitrary amount of code can be emitted inline so flexibility is not removed.

Reply:
Wow thanks I understand how the hash works now. I thought it was was the entire node